podpac/datalib/gfs.py

```diff
@@ -0,0 +1,126 @@
+bucket = s3.Bucket(BUCKET)
+
+# TODO add time to native_coordinates
+class GFSSource(Rasterio):
+    parameter = tl.Unicode(readonly=True).tag(attr=True)
+    level = tl.Unicode(readonly=True).tag(attr=True)
+    date = tl.Unicode(readonly=True).tag(attr=True)
+    hour = tl.Unicode(readonly=True).tag(attr=True)
+    forecast = tl.Unicode(readonly=True).tag(attr=True)
+    dataset = tl.Any(readonly=True)
+
+    def init(self):
+        self._logger = logging.getLogger(__name__)
+        
+        # check if the key exists
+        try:
+            s3.Object(BUCKET, self._key).load()
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                raise ValueError("Not found: '%s'" % self._key) # TODO list options
+            else:
+                raise
+    
+    @property
+    def _key(self):
+        return '%s/%s/%s/%s/%s' % (self.parameter, self.level, self.date, self.hour, self.forecast)
+
+    @tl.default('nan_vals')
+    def _get_nan_vals(self):
+        return [self.dataset.nodata]
+        # return list(self.dataset.nodatavals) # which?
+
+    @property
+    def source(self):
+        return self._key
+
+    @tl.default('dataset')
+    def open_dataset(self):
+        """Opens the data source"""
+
+        cache_key = 'fileobj'
+        with rasterio.MemoryFile() as f:
+            if self.cache_ctrl and self.has_cache(key=cache_key):
+                data = self.get_cache(key=cache_key)
+                f.write(data)
+            else:
+                self._logger.info('Downloading S3 fileobj (Bucket: %s, Key: %s)' % (BUCKET, self._key))
+                s3.Object(BUCKET, self._key).download_fileobj(f)
+                f.seek(0)
+                self.cache_ctrl and self.put_cache(f.read(), key=cache_key)
+            
+            dataset = f.open()
+
+        return dataset
+
+class GFS(DataSource):
+    parameter = tl.Unicode(readonly=True).tag(attr=True)
+    level = tl.Unicode(readonly=True).tag(attr=True)
+    date = tl.Unicode(readonly=True).tag(attr=True)
+    hour = tl.Unicode(readonly=True).tag(attr=True)
+
+    @property
+    def source(self):
+        return '%s/%s/%s/%s' % (self.parameter, self.level, self.date, self.hour)
+
+    def init(self):
+        # TODO check prefix and the options at the next level
+        
+        self._prefix = '%s/%s/%s/%s/' % (self.parameter, self.level, self.date, self.hour)
+        self.forecasts = [obj.key.replace(self._prefix, '') for obj in bucket.objects.filter(Prefix=self._prefix)]
+        
+        if not self.forecasts:
+            raise ValueError("Not found: '%s/*'" % self._prefix)
+
+        params = {'parameter': self.parameter, 'level': self.level, 'date': self.date, 'hour': self.hour, 'cache_ctrl': self.cache_ctrl}
+        self._sources = np.array([GFSSource(forecast=h, **params) for h in self.forecasts]) # can we load this lazily?
```

Comment:
Oh, yeah, but that's not what I mean. I mean that I want to create the individual elements lazily. Regardless of whether I use a traitlets default or the init, I want to make the whole array at that time, but only have it create the individual GFSSource nodes when they are needed for a particular evaluation (which probably won't need all of them). Creating a GFSSource requires an S3 request, so it would be a slight an optimization.

BTW, At first I tried to use a Compositor for the GFS node, but the Compositor is pretty complicated, and I wasn't sure it would do what I wanted anyway. Would the Compositor interpolate times? Each source is whole-world data for a particular time.

Reply:
Ah! That makes sens. I guess you could do you iteroutputs trick? Maybe? Or maybe make the GFSSource delay the S3 request until requested? 

Yeah, the compositor needs work. It should at some point be able to handle time interpolation, but it doesn't do that right now. I'm not sure what the best path forward is. It may be that a compositor is a type of DataSource... I need to make a bunch of issues to keep track of all these little thing.

Reply:
Yes, it would make a lot of sense if the Compositor was a DataSource. The `native_coordinates` would be tricky. In this case (and it is not handled at all), the sources change to a lower spatial resolution in the later forecasts. So `GFS` has a bug in the native_coordinates. I chose the earlier, higher resolution spatial coordinates, but really I need something like a group with two Coordinates (and the ability to interpolate in the space between them).

Reply:
Hmmm... right, the native coordinates is why they are split. Perhaps then we need an interpolation node, then branch off into datasource and compositor... so that they can use some of the same machinery but not all... worth thinking about. Glad  you and @mlshapiro  now have experience with these problems. :)